Step2

goyounggyeong opened the pull request:
2단계 - 문자열 계산기
코드리뷰 부탁드립니다

gogoyg committed:
study: kotlin 기본 학습

gogoyg committed:
2단계 - 문자열 계산기

Rok93 commented on build.gradle.kts:
```diff
@@ -13,7 +13,7 @@ repositories {
     testImplementation("org.junit.jupiter", "junit-jupiter", "5.8.2")
     testImplementation("org.assertj", "assertj-core", "3.22.0")
-    testImplementation("io.kotest", "kotest-runner-junit5", "5.2.3")
+    testImplementation("io.kotest", "kotest-runner-junit5", "5.5.0")
```
Kotest 5.5.0에서는 [테스트코드가 스킵되는 버그](https://kotest.io/docs/changelog.html#551-october-2022)가 있었던 것으로 파악돼요. 
5.5.1 버전으로 변경해보면 어떨까요? 😃 

![image](https://github.com/next-step/kotlin-racingcar/assets/47850258/f7e06dd2-79bb-46d5-9949-09d9d9acfa80)


![image](https://github.com/next-step/kotlin-racingcar/assets/47850258/3a9dbcc3-0f45-4966-b28e-ac535d269c6f)

Rok93 commented on src/test/kotlin/study/PersonTest.kt:
```diff
@@ -0,0 +1,32 @@
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.Test
+
+class PersonTest {
```
1단계 미션을 수행하면서 커밋했던 이력이 2단계에서도 동일하게 반영되어있네요. 
아마도 `git rebase`를 수행하지 않으셨거나 혹은 git rebase가 제대로 수행되지 않은 것으로 파악돼요. 

[온라인 코드리뷰 요청 3단계](https://github.com/next-step/nextstep-docs/blob/master/codereview/review-step3.md)를 참고해서 git rebase를 한번 진행해주시면 좋을 것 같아요. 🙏

Rok93 replied:
아직 rebase가 안된 것 같아요. 🥲

<img width="1309" alt="image" src="https://github.com/next-step/kotlin-racingcar/assets/47850258/7d544182-5d2c-4365-9c6b-d1bcb2f2cad4">

이번 단계는 이전 스텝의 커밋이 끼여있어도, 작업한 내용을 보는데 그렇게 불편하지 않으니 다음 스텝에는 꼭 git rebase 부탁드려요! 🙏

Rok93 commented on src/main/kotlin/calculator/calculateFromString.kt:
```diff
@@ -0,0 +1,17 @@
+package calculator
+
+fun calculateFromString(expression: String): Int {
```
[코틀린 코드 컨벤션 문서](https://kotlinlang.org/docs/coding-conventions.html#source-file-names)에 따르면 해당 top-level function을 가지는 Kotlin 파일의 명칭은, `upper camel case`로 작성하라고 가이드하고있어요. 😃 

컨벤션에 맞춰서 변경해보면 어떨까요?

Rok93 replied:
문자열 계산기 도메인의 핵심 기능을 `최상위 함수`로 뽑아서 쓰기보다는 도메인 객체로 관리하고 처리하도록 변경해보면 어떨까요? 

개인적인 생각으로 `문자열 계산기`는 일반적인 계산기와 다르게 `문자열 수식`(ex. 1 + 3 * 10 / 5)을 입력받으면 수식을 계산한 값을 반환해줘야합니다. 하지만 현재 Calculator 객체는  `피연산자1`, `연산자`, `피연산자2`를 주입받아서 단순 계산을 하고있는, 일반적인 계산기와 다를바가 없어요. 

제가 생각하기에는 `문자열 계산기`라는 비즈니스 로직에 특화된 `문자열 계산기 객체`는 아래와 같은 구조를 띄는 것이 조금 더 좋지 않을까 싶어요. 🤔 (현재는 이 문자열 계산 이라는 가장 핵심 로직이 최상위 함수인 `calculateFromString`이라는 함수로 구현이 되어있네요. 👀)

```kotlin
class StringCalculator(
    ....
) {
    fun calculate(expression: String): Int {
        // 생략
    }
}

Rok93 commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -0,0 +1,17 @@
+    }
+
+    private fun add() = firstElement + secondElement
+    private fun subtract() = firstElement + secondElement
```
뺄셈 기능이 잘못 구현된 것 같아요. 👀
만약 테스트 코드가 있었다면, 이 실수를 빠르게 감지할 수 있었을거에요. 😃
테스트 코드로 +, -, *, / `각 연산에 대한 테스트 코드`를 추가로 작성해주세요. 😊

Rok93 replied:
add(), subtract(), multiply(), divide()를 굳이 메서드로 쪼갤 필요없이, `when` 구문 안에서 바로 구현해도 되지 않을까요? 🤔

```kotlin
fun calcuate() = when (operator) {
        '+' -> firstElement + secondElement
        // 생략
}
```

goyounggyeong replied:
inline으로 변경했습니다

Rok93 commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -0,0 +1,17 @@
+    private fun add() = firstElement + secondElement
+    private fun subtract() = firstElement + secondElement
+    private fun multiply() = firstElement * secondElement
+    private fun divide() = firstElement / secondElement
```
나누기 연산 시, 만약 secondElement가 0인 경우에는 어떤 문제가 발생할까요? 🤔

Rok93 commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -0,0 +1,17 @@
+
+class Calculator(private val firstElement: Int, private val operator: Char, private val secondElement: Int) {
+
+    fun calcuate() = when (operator) {
```
아마도 오타인 것 같아요. 🤔
~~calcuate~~ ->  **calculate**

Rok93 replied:
when 구문 활용 👍👍👍 

추가적으로 메서드의 `반환타입`을 명시한 케이스와 명시하지 않은 케이스가 있는데, 가급적 통일해서 코드를 작성하는 것이 좋을 것 같아요. 👀 

저는 개인적으로 변수의 경우에는 `타입추론`을 적극적으로 활용하는 편이고, 메서드의 경우 반환타입을 꼭 명시해주는 편이에요!

goyounggyeong replied:
차후 과제에 반영하겠습니다.

Rok93 commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -0,0 +1,17 @@
+        '-' -> subtract()
+        '*' -> multiply()
+        '/' -> divide()
+        else -> throw IllegalArgumentException(" 사칙연산기호 확인 ")
```
예외가 발생했을 때, `operator`로 어떤 문자가 들어왔는지 예외 메시지에 포함시켜서 보여주면 디버깅이 더 쉽지 않을까요? 😃

goyounggyeong replied:
변경했습니다

Rok93 commented on src/main/kotlin/calculator/calculateFromString.kt:
```diff
@@ -0,0 +1,17 @@
+    val expressionQueue = buildExpressionQueue(expression.replace(" ", ""))
+    var result = 0
+    var operator = '+'
+    while (!expressionQueue.isEmpty()) {
```
아래와 같이 `isNotEmpty()` 라는 확장 함수로 변경해볼 수 있을 것 같아요. 😃
```suggestion
    while (expressionQueue.isNotEmpty()) {
```

Rok93 commented on src/main/kotlin/calculator/calculateFromString.kt:
```diff
@@ -0,0 +1,17 @@
+package calculator
+
+fun calculateFromString(expression: String): Int {
+    val expressionQueue = buildExpressionQueue(expression.replace(" ", ""))
```
별도의 (최상위) 함수 구현 없이, String의 확장함수인 `split`으로 구분자로 수식을 분리하고 필요에따라 피연산자들은 숫자로 변환, 연산자들은 연산자의 타입으로 변경해보면 어떨까요? 🤔

```suggestion
    val expressionQueue = expression.split(" ") // 반환타입 List<String> 
        .toCollection(ArrayDeque<Any>()) // 반환타입 ArrayDeque<Any> 
```

또한 현재 Queue의 제네릭 타입이 `Any`인데, 이는 Java에서 Object와 동일합니다. 
가급적 Object 타입을 직접적으로 다루는 일은 지양하는 것이 좋기 때문에, 수식에서 연산자와 피연사를 하나의 컬렉션에 모아서 뽑는 것이 아니라, 분리해서 연산자 컬렉션, 피연산자 컬렉션으로 분리하여 추출한다면, 하나의 Any 타입 사용 없이도 구현이 가능하여, 타입으로부터 보호받는 더 안전한 프로그래밍이 가능해질 것 같아요. 😃

수식은 `짝수번째 위치에는 피연산자`, `홀수번째 위치에는 연산자`가 있기 때문에 이를 적절히 이용하면 구현이 그렇게 어렵지 않을 것이라 생각됩니다. 😉

goyounggyeong replied:
문제 내용에 기호와 숫자의 split 기준이 적혀있지 않았던 것 같습니다. 그래서 고민하다 "33/ 3* 3" 의 경우와 같이 띄어쓰기가 지켜지지 않을 경우를 위해 split(" ")를 사용하지 않았습니다.
Any 사용은 지양하도록 하겠습니다.
리펙토링후 반영했습니다.

Rok93 replied:
문자열계산기 [기능요구사항](https://edu.nextstep.camp/s/w5G1pCe7/ls/fo3Uc7Yc)에 따르면 
> 예를 들어 "2 + 3 * 4 / 2"와 같은 문자열을 입력할 경우 2 + 3 * 4 / 2 실행 결과인 10을 출력해야 한다.

저는 위의 문구에서 구분자가 `" "`라고 생각했었는데, 직접적으로 명시된건 아니라 혼동의 여지가 조금 있었던 것 같네요. 😃

Rok93 reviewed (changes requested):
영경님 문자열 계산기 구현 잘해주셨네요. 👍👍 

구현하신 코드 살펴보면서 짚고 넘어가면 좋을 것 같은 부분이 있어서, 몇몇 코멘트 남겨두었습니다. 😃
확인해서 반영해주셨으면합니다. 😊

조금 더 첨언을드리자면...
현재 대다수의 핵심 비즈니스 로직이 `top-level function`에 구현되어있는 것 같은데, 
이를 도메인 객체가 비즈니스 로직을 수행할 수 있도록 변경해보셨으면 좋겠어요. 😉
혹시라도 피드백 반영하면서 이해가 잘 안되거나 어려운 부분 있으면 언제든지 편하게 DM 주세요! 🙏

gogoyg committed:
코드리뷰 반영
- top-level function 제거
- junit version 변경
- 뺄셈기능 오류 해소 및 계산자 inline 처리
- 오타해소
- Exception 안내문구추가
- not keyword 제거
- Any 사용 지양

goyounggyeong commented:
안녕하세요. 코맨트 해주신 내용 반영했습니다.
꼼꼼한 리뷰 감사합니다.

Rok93 commented on src/test/kotlin/study/StringTest.kt:
```diff
@@ -0,0 +1,12 @@
+        assertThat("".isEmpty()).isTrue()
+        assertThat("a".isEmpty()).isFalse()
+    }
+}
```
현재 Ktlint Check를 돌려보면 StringTest 클래스의 가장 마지막 줄이 new Line(= \n)이 아니라서 통과하지 못하는 것을 알 수 있어요. 

<img width="1470" alt="image" src="https://github.com/next-step/kotlin-racingcar/assets/47850258/62f29157-4811-4434-9663-c8d8cfc3bc1b">


이번 기회에, Ktlint 설정을 추가해서 Ktlint Check를 강제하도록 설정해보면 어떨까요? 🤗 

```gradle
./gradlew ktlintApplyToIdea
mkdir .git/hooks
./gradlew addKtlintCheckGitPreCommitHook
./gradlew clean check
```

Rok93 replied:
추가적으로 커밋 메시지를 [프로그래밍 요구 사항](https://edu.nextstep.camp/s/w5G1pCe7/ls/fdzp3quC)에 따라 [AngularJS Git Commit Message Conventions](https://gist.github.com/stephenparish/9941e89d80e2bc58a153)에 맞춰서 작성해주시면 좋을 것 같아요. 😃

5a1174481756e8dc459c9d88869fd60d9a1de45b

Rok93 commented on src/main/kotlin/calculator/StringCalculator.kt:
```diff
@@ -0,0 +1,43 @@
+        return result
+    }
+
+    fun isNumeric(toCheck: String): Boolean {
```
StringCalculator 클래스 내부에서만 쓰이는 메서드인것 같아요. 👀
접근제한자를 `private`으로 변경하는게 좋을 것 같아요. 😃

Rok93 replied:
추가적으로 메서드 파라미터명을 조금 더 의미있는 네이밍으로 변경해보면 어떨까요? 😃

Rok93 commented on src/main/kotlin/calculator/StringCalculator.kt:
```diff
@@ -0,0 +1,43 @@
+        return toCheck.matches(regex)
+    }
+
+    fun buildExpressionQueue(sanitizedExpression: String): Queue<String> {
```
`isNumeric`과 `buildExpressionQueue` 메서드 모두 calculate 메서드 내부에서 사용되고 있어요. 

코드를 읽는 순서를 고려했을 때, 메서드를 메서드가 호출되는 순서대로 아래와 같이 배치하면 조금 더 가독성이 좋아지지 않을까요? 😄

1. buildExpressionQueue
2. isNumeric

Rok93 commented on src/main/kotlin/calculator/StringCalculator.kt:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    fun isNumeric(toCheck: String): Boolean {
+        val regex = "-?[0-9]+(\\.[0-9]+)?".toRegex()
```
숫자인지 검증을 할 때마다 매번 동일한 `Regex 객체`가 생성될 것 같아요. 🤔
해당 정규식을 상수로 뽑아서 캐싱해서 사용해보면 어떨까요? 😃

[아이템6 불필요한 객체 생성을 피하라](https://incheol-jung.gitbook.io/docs/study/effective-java/undefined/2020-03-20-effective-6item)

Rok93 commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -0,0 +1,14 @@
+package calculator
+
+class Calculator(private val firstElement: Int, private val operator: String, private val secondElement: Int) {
+    fun calculate() = when (operator) {
```
연산(덧셈, 뺄셈, 곱셈, 나눗셈)을 하는 역할은 계산기보다는 `연산자` 객체가 `operate`하는 것으로 볼 수 있을 것 같아요. 🤔
Calculator 대신 `Operator` 객체로 변경하고, operate 메서드로 변경해보면 어떨까요? 😃

Rok93 commented on src/main/kotlin/calculator/StringCalculator.kt:
```diff
@@ -0,0 +1,43 @@
+        var currentNumber = ""
+        for (char in sanitizedExpression) {
+            currentNumber = if (char.isDigit()) {
+                currentNumber + "" + char
```
아래와 같이 `""` 를 제거해도 동일하게 동작할 것 같아요. 😃
```suggestion
                currentNumber + char
```

Rok93 reviewed (approved):
영경님 피드백 잘 반영해주셨네요. 😃👍
추가적으로 몇몇 코멘트 더 남겨두었으니 3단계 미션 시작하기전에 같이 반영해주시면 좋을 것 같아요. 🙏

다음 스텝은 꼭 `git rebase` 부탁드려요!
미션 수행하시면서 어려운 점이 있으시면 언제든지 DM주세요. 😉
